rust/chains/abacus-ethereum/src/interchain_gas.rs

```diff
@@ -0,0 +1,192 @@
+#![allow(missing_docs)]
+
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use ethers::contract::abigen;
+use ethers::prelude::*;
+use eyre::Result;
+use tracing::instrument;
+
+use abacus_core::{
+    AbacusContract, ContractLocator, Indexer, InterchainGasPaymaster,
+    InterchainGasPaymasterIndexer, InterchainGasPayment,
+};
+
+use crate::trait_builder::MakeableWithProvider;
+
+abigen!(
+    EthereumInterchainGasPaymasterInternal,
+    "./chains/abacus-ethereum/abis/InterchainGasPaymaster.abi.json"
+);
+
+impl<M> std::fmt::Display for EthereumInterchainGasPaymasterInternal<M>
+where
+    M: Middleware,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+
+pub struct InterchainGasPaymasterIndexerBuilder {
+    pub from_height: u32,
+    pub chunk_size: u32,
+}
+
+impl MakeableWithProvider for InterchainGasPaymasterIndexerBuilder {
+    type Output = Box<dyn InterchainGasPaymasterIndexer>;
+
+    fn make_with_provider<M: Middleware + 'static>(
+        &self,
+        provider: M,
+        locator: &ContractLocator,
+    ) -> Self::Output {
+        Box::new(EthereumInterchainGasPaymasterIndexer::new(
+            Arc::new(provider),
+            locator,
+            self.from_height,
+            self.chunk_size,
+        ))
+    }
+}
+
+#[derive(Debug)]
+/// Struct that retrieves event data for an Ethereum InterchainGasPaymaster
+pub struct EthereumInterchainGasPaymasterIndexer<M>
+where
+    M: Middleware,
+{
+    contract: Arc<EthereumInterchainGasPaymasterInternal<M>>,
+    provider: Arc<M>,
+    #[allow(unused)]
+    from_height: u32,
+    #[allow(unused)]
+    chunk_size: u32,
+}
+
+impl<M> EthereumInterchainGasPaymasterIndexer<M>
+where
+    M: Middleware + 'static,
+{
+    /// Create new EthereumInterchainGasPaymasterIndexer
+    pub fn new(
+        provider: Arc<M>,
```

Comment:
Does the provider need to be an `Arc<M>`? Should be fine to just be `M`

Reply:
the ethers contract expects an `Arc<M>` https://docs.rs/ethers/latest/ethers/contract/struct.Contract.html#method.new